packages/flutter/lib/src/widgets/splash.dart

```diff
@@ -0,0 +1,621 @@
+  }
+
+  /// Computes the [Matrix4] that allows [fromRenderObject] to perform paint
+  /// in [toRenderObject]'s coordinate space.
+  ///
+  /// Typically, this is used to find the transformation to apply to the [controller]
+  /// so it matches the [referenceBox].
+  ///
+  /// Returns null if either [fromRenderObject] or [toRenderObject] is not
+  /// in the same render tree, or either of them is in an offscreen subtree
+  /// (see [RenderObject.paintsChild]).
+  static Matrix4? getPaintTransform(
+    RenderObject fromRenderObject,
+    RenderObject toRenderObject,
+  ) {
+    // The paths to fromRenderObject and toRenderObject's common ancestor.
+    final List<RenderObject> fromPath = <RenderObject>[fromRenderObject];
+    final List<RenderObject> toPath = <RenderObject>[toRenderObject];
+
+    RenderObject from = fromRenderObject;
+    RenderObject to = toRenderObject;
+
+    while (!identical(from, to)) {
+      final int fromDepth = from.depth;
+      final int toDepth = to.depth;
+
+      if (fromDepth >= toDepth) {
+        final RenderObject? fromParent = from.parent;
+        // Return early if the 2 render objects are not in the same render tree,
+        // or either of them is offscreen and thus won't get painted.
+        if (fromParent is! RenderObject || !fromParent.paintsChild(from)) {
+          return null;
+        }
+        fromPath.add(fromParent);
+        from = fromParent;
+      }
+
+      if (fromDepth <= toDepth) {
+        final RenderObject? toParent = to.parent;
+        if (toParent is! RenderObject || !toParent.paintsChild(to)) {
+          return null;
+        }
+        toPath.add(toParent);
+        to = toParent;
+      }
+    }
+    assert(identical(from, to));
+
+    final Matrix4 transform = Matrix4.identity();
+    final Matrix4 inverseTransform = Matrix4.identity();
+
+    for (int index = toPath.length - 1; index > 0; index -= 1) {
+      toPath[index].applyPaintTransform(toPath[index - 1], transform);
+    }
+    for (int index = fromPath.length - 1; index > 0; index -= 1) {
+      fromPath[index].applyPaintTransform(fromPath[index - 1], inverseTransform);
+    }
+
+    final double det = inverseTransform.invert();
+    return det != 0 ? (inverseTransform..multiply(transform)) : null;
+  }
+
+  /// Called when the user input that triggered this feature's appearance was confirmed.
+  ///
+  /// Typically causes the [Splash] to propagate faster across the surface.
+  /// By default this method does nothing.
+  void confirm() {}
+
+  /// Called when the user input that triggered this feature's appearance was canceled.
+  ///
+  /// Typically causes the [Splash] to gradually disappear.
+  /// By default this method does nothing.
+  void cancel() {}
+
+  /// {@macro flutter.widgets.splash.SplashController}
+  final SplashController controller;
+
+  /// The render box whose visual position defines the splash effect's
+  /// frame of reference.
+  final RenderBox referenceBox;
```

Comment:
We probably want it to be a `RenderObject referenceObject`, not specifically the box protocol.

(I wonder if it disrupts anything in the implementation.)

Reply:
Non-box shapes still use `RenderBox` objects, so the primary reason to switch to `RenderObject` would be to allow either a [`RenderBox`](https://api.flutter.dev/flutter/rendering/RenderBox-class.html) or a [`RenderSliver`](https://api.flutter.dev/flutter/rendering/RenderSliver-class.html) to be passed.

I have very little experience with render object classes, but it looks like making this change would break the current implementations, since there isn't a `size` getter for `RenderObject` / `RenderSliver`.

Reply:
Where is the `size` used? The only place I found that uses the `size` is `_RenderInkFeatures.paint`, which uses the size to clip:
```
      canvas.clipRect(Offset.zero & size);
```

It show that, rendering splashes doesn't need sizes or the box protocol at all. It's the clipping that requires the size (which makes sense, since it's a box clip).

Reply:
I believe that the `size` is needed for any `Splash` effect that expands to fill the widget it's being painted on, including the Material design splash factories as well as whatever expanding splashes we create in the future.

Aside from `RenderBox`, is there another type of render object that we would want to use for splashes?

Reply:
I'm thinking of creating a more general widget that contains protocol-agnostic logic, such as managing splashes. But on a second thought it's ok to start from box and don't overplan at least for now.